c7n/resources/secretsmanager.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2018 Capital One Services, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from botocore.exceptions import ClientError
+
+from c7n.manager import resources
+from c7n.filters import FilterRegistry
+from c7n.query import QueryResourceManager
+from c7n.tags import RemoveTag, Tag, TagActionFilter, TagDelayedAction
+from c7n.utils import local_session
+
+
+@resources.register('secrets-manager')
+class SecretsManager(QueryResourceManager):
+    filters = FilterRegistry('secrets-manager.filters')
+    filters.register('marked-for-op', TagActionFilter)
+    filter_registry = filters
+    permissions = ('secretsmanager:ListSecretVersionIds',)
+
+    class resource_type(object):
+        service = 'secretsmanager'
+        enum_spec = ('list_secrets', 'SecretList', None)
+        detail_spec = ('describe_secret', 'SecretId', 'ARN', None)
+        id = 'ARN'
+        name = 'Name'
+        dimension = None
+        filter_name = None
+
+    def augment(self, resources):
+        client = local_session(self.session_factory).client('secretsmanager')
+
+        def _augment(r):
+            r['Versions'] = client.list_secret_version_ids(SecretId=r['ARN']).get('Versions')
+            self.log.info('adding version info')
+            return r
+
+        resources = super(SecretsManager, self).augment(resources)
+
+        with self.executor_factory(max_workers=1) as w:
+            return list(filter(None, w.map(_augment, resources)))
+
+
+@SecretsManager.action_registry.register('tag')
+class TagSecretsManagerResource(Tag):
+    """Action to create tag(s) on a Secret resource
+
+    :example:
+
+    .. code-block:: yaml
+
+        policies:
+            - name: tag-secret
+              resource: secrets-manager
+              actions:
+                - type: tag
+                  key: tag-key
+                  value: tag-value
+    """
+
+    permissions = ('secretsmanager:TagResource',)
+
+    def process_resource_set(self, resources, tags):
+        client = local_session(self.manager.session_factory).client('secretsmanager')
+        for r in resources:
+            r['Tags'].extend([{'Key': t['Key'], 'Value':t['Value']} for t in tags])
+            try:
+                client.tag_resource(SecretId=r['ARN'], Tags=r['Tags'])
+            except ClientError as e:
+                if e['Error']['Code'] == 'InternalFailure':
```

Comment:
internal failures should be re-raising.

Reply:
This is a result of a bug with tag name conflicts it _should_ be a client error, snippet to reproduce:

```python
client = boto3.client('secretsmanager')
client.tag_resource(SecretId="test", Tags=[{"Key":"test", "Value":"value"},{"Key":"test", "Value":"value"}])

a ticket's been open with AWS for now

Reply:
yeah, for now i suggest we not catch here then, the internal failure should hopefully get fixed by the service team. ie. those errors need human attention. switching how you pass the tags should also resolve.